NExtends/Expressions/ExpressionReducer.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace NExtends.Expressions
+{
+	public class ExpressionReducer<TSource, TProp> : ExpressionVisitor
+	{
+		MemberExpression Reducer { get; set; }
+		ParameterExpression Parameter { get; set; }
+
+		bool CurrentNodeDependsOnPreviousParameter { get; set; }
+
+		protected ExpressionReducer(MemberExpression reducer, ParameterExpression parameter)
+		{
+			Parameter = parameter;
+			Reducer = reducer;
+
+			if (Reducer.Member.DeclaringType != typeof(TSource)
+				&& Reducer.Member.DeclaringType.IsAssignableFrom(typeof(TSource)))
+			{
+				PropertyInfo targetProperty;
+				if (Reducer.Member.DeclaringType.IsInterface)
+				{
+					targetProperty = GetClassProperty(Reducer.Member.DeclaringType, typeof(TSource), Reducer.Member.Name);
+				}
+				else
+				{
+					targetProperty = typeof(TSource).GetProperty(Reducer.Member.Name);
+				}
+				Reducer = Expression.Property(Reducer.Expression as ParameterExpression, targetProperty);
+			}
+		}
+
+		static PropertyInfo GetClassProperty(Type interfaceType, Type classType, string propertyName)
+		{
+			var nameProperty = interfaceType.GetProperty(propertyName);
+			var mapping = classType.GetInterfaceMap(interfaceType);
+			var nameGetter = nameProperty.GetGetMethod();
+
+			MethodInfo targetMethod = null;
+			for (var i = 0; i < mapping.InterfaceMethods.Length; i++)
+			{
+				if (mapping.InterfaceMethods[i] == nameGetter)
+				{
+					targetMethod = mapping.TargetMethods[i];
+					break;
+				}
+			}
+
+			PropertyInfo targetProperty = null;
+			foreach (var property in classType.GetProperties(BindingFlags.Instance | BindingFlags.GetProperty | BindingFlags.Public | BindingFlags.NonPublic))
+			{
+				if (targetMethod == property.GetGetMethod(true))
+				{
+					targetProperty = property;
+					break;
+				}
+			}
+
+			return targetProperty;
+		}
+
+		public static Expression<Func<TProp, bool>> Reduce(Expression<Func<TSource, bool>> expression, Expression<Func<TSource, TProp>> reducer, bool defaultValue)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'System.Linq.Expressions.LambdaExpression' instead of 'System.Linq.Expressions.Expression<System.Func<TSource, bool>>'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3242)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'System.Linq.Expressions.LambdaExpression' instead of 'System.Linq.Expressions.Expression<System.Func<TSource, TProp>>'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3242)